### PR TITLE
Fourier-Chebyshev Fit to `B` for particle tracing

### PR DESCRIPTION
**Pull request by YigitElma**
- Adds a new convenience class `FourierChebyshevField` which has a couple of methods to be compatible with optimization,
  - `init` : only resolution info
  - `build` : creates grid and transforms
  - `fit` : given params and profiles (these change during optimization), fits a new Fourier-Chebyshev series. 
  - `evaluate` : Evaluate the series at requested point.

Example usage will be,
```python
interpolater = FourierChebyshevField(L=eq.L_grid, M=eq.M_grid, N=eq.N_grid)
interpolater.build(eq)
data_fit = interpolater.fit(eq.params_dict, {"iota": eq.iota})

rtz, _, aux = _trace_particles(
    field=interpolator,  # normally you pass eq here
    y0=x0,
    model=model,
    model_args=args,
    ts=ts,
    params=None,   # normally you pass eq.params_dict here
    max_steps=int((ts[-1] - ts[0]) / 1e-8),
    min_step_size=1e-8,
    stepsize_controller=PIDController(rtol=1e-4, atol=1e-4, dtmin=1e-8),
    saveat=SaveAt(ts=ts),
    solver=Tsit5(),
    adjoint=RecursiveCheckpointAdjoint(),
    event=event,
    chunk_size=None,
    options={},    # this passes extra stuff to field computation, not needed here
    throw=False,
    return_aux=True,
)
```

Todo:
- [ ] `trace_particles` should initialize `FourierChebyshevField` and by default use it. Add a flag to use Equilibrium field
- [ ] `plot_particle_trajectories` etc should also be able to use the interpolated field.
- [ ] Look at the spectral convergence (ideally lower res for basis vectors might be good, but current code is simpler and with batching might not be inefficient)
- [ ] Consider adding spline version too?

### CODE REVIEW

File: desc/particles.py [hunk 1]
```diff
@@ -1242,3 +1281,195 @@ def _intfun_wrapper(
...
+        eq : Equilibrium
+            Equilibrium to be used to get transforms.
+
+        """
+        self.data_keys = ["B", "grad(|B|)", "e^rho", "e^theta", "e^zeta"]
+        self.l = jnp.arange(self.L)
+        self.M_fft = 2 * self.M + 1
+        self.N_fft = 2 * self.N + 1
+        self.m = jnp.fft.fftfreq(self.M_fft) * self.M_fft
+        self.n = jnp.fft.fftfreq(self.N_fft) * self.N_fft
```
Comment (f0uriest) [inline desc/particles.py:1332]: i think this is missing a factor of NFP

File: desc/particles.py [hunk 2]
```diff
@@ -1242,3 +1281,195 @@ def _intfun_wrapper(
...
+        can be used to get the fit faster. This method creates the grid and transforms
+        to be used during the fitting procedure.
+
+        Parameters
+        ----------
+        eq : Equilibrium
+            Equilibrium to be used to get transforms.
+
+        """
+        self.data_keys = ["B", "grad(|B|)", "e^rho", "e^theta", "e^zeta"]
```
Comment (f0uriest) [inline desc/particles.py]: `e^theta` diverges on axis so it can't be fit well, you can probably use `e^theta*sqrt(g)`

Comment (f0uriest) [reply desc/particles.py]: also just to note, `e^rho` and `e^theta*sqrt(g)` don't scale like rho^m near axis so shouldn't be fit to a zernike basis

File: desc/particles.py [hunk 3]
```diff
@@ -1242,3 +1281,195 @@ def _intfun_wrapper(
...
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self.data_keys,
+            params,
+            self.transforms,
+            profiles,
+        )
+        rho = self.grid.nodes[:, 0]
+        L, M, N = self.L, self.M_fft, self.N_fft
+        data_raw_shaped = {
+            "Br": data_raw["B"][:, 0].reshape(N, L, M),
```
Comment (f0uriest) [inline desc/particles.py]: i think the natural order should be M, L, N with order="F"? but either way its safer to use `grid.meshgrid_reshape` here

Comment (f0uriest) [reply desc/particles.py]: also might be more efficient to stack everything into one array and then use vmap etc

**Review by f0uriest — COMMENTED** [review]
from an API standpoint, I'd prefer if you pass in the FourierChebyshevField as the field object rather than a dict of params, and pass the params separately the same way we do for an equilibrium (and allow the FourierChebyshevField to contain the params as an attribute)

File: desc/particles.py (hunk 2 above)
Comment (YigitElma) [reply desc/particles.py]: For `e^theta`, I fit the series to `e^theta*rho`, which is used in guiding center equations (so, I guess easier than `sqrt(g)`). It was also behaving well.

File: desc/particles.py (hunk 1 above)
Comment (YigitElma) [reply desc/particles.py:1332]: Yes, the current grid selection is not the best. I create a grid with NFP=1 instead of equilibriums NFP.

File: desc/particles.py [hunk 4]
```diff
@@ -1246,3 +1283,238 @@ def _intfun_wrapper(
...
+    """
+
+    _static_attrs = ["L", "M", "N", "M_fft", "N_fft", "data_keys"]
+
+    def __init__(self, L, M, N):
+        self.L = L
+        self.M = M
+        self.N = N
+
+    def build(self, eq):
```
Comment (YigitElma) [inline desc/particles.py:1314]: It would be actually nice to be able to flexibly compute things in this class. Like different trajectory models might need different keys. Probably, having some options would work.

File: desc/particles.py [hunk 5]
```diff
@@ -1246,3 +1283,238 @@ def _intfun_wrapper(
...
+        data_raw = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self.data_keys,
+            params,
+            self.transforms,
+            profiles,
+        )
+        rho = self.grid.nodes[:, 0]
+        L, M, N = self.L, self.M_fft, self.N_fft
+        keys = [
```
Comment (YigitElma) [inline desc/particles.py]: This can be automated for different keys as,
```python
keys = [key + i for key in self.data_keys for i in ["_r", "_p", "_z"]]
```

File: desc/particles.py [hunk 6]
```diff
@@ -1246,3 +1283,238 @@ def _intfun_wrapper(
...
+        # grad(|B|)
+        gradB = results[3:6]
+        # e^rho
+        er = results[6:9]
+        # e^theta*rho
+        et_x_rho = results[9:12]
+        # e^zeta
+        ez = results[12:15]
+
+        return b, magB, gradB, er, et_x_rho, ez
```
Comment (YigitElma) [inline desc/particles.py]: Cleaner way is to return a dict with same keys as `self.data_keys`

**Comment by YigitElma** [comment]
Not sure if this is necessary but we can reduce the number of fits by directly fitting the dot products. This will probably require higher fit resolution (since dot product increases the frequency window) but it may be more efficient. Currently, I fit 15 quantities, with the dot product you need 7 fits.

<details>
<summary>Here is the code implementation</summary>

```python
class VacuumGuidingCenterTrajectoryOpt(VacuumGuidingCenterTrajectory):
    @jit
    def vf(self, t, x, args):
        x = x.squeeze()
        model_args, field, params, kwargs = args
        m, q, mu = model_args
        xp, yp, zeta, vpar = x
        rho = jnp.sqrt(xp**2 + yp**2)
        theta = jnp.arctan2(yp, xp)
        # compute functions are not correct for very small rho
        rho = jnp.where(rho < 1e-6, 1e-6, rho)

        data = field.evaluate(rho, theta, zeta)

        # take dot product for rho, theta and zeta coordinates
        common = (m / q / data["|B|"] ** 2) * ((mu * data["|B|"] / m) + vpar**2)
        rhodot = common * data["(bxgrad(|B|))*e^rho"] # b . e^rho = 0
        thetadot_x_rho = vpar * data["b*e^theta*rho"] + common * data["(bxgrad(|B|))*e^theta*rho"]
        zetadot = vpar * data["b*e^zeta"] + common * data["(bxgrad(|B|))*e^zeta"]

        # get the derivative for cartesian-like coordinates
        xpdot = rhodot * jnp.cos(theta) - thetadot_x_rho * jnp.sin(theta)
        ypdot = rhodot * jnp.sin(theta) + thetadot_x_rho * jnp.cos(theta)
        # derivative the parallel velocity
        vpardot = -mu / m * data["b*grad(|B|)"]
        dxdt = jnp.array([xpdot, ypdot, zetadot, vpardot]).reshape(x.shape)
        return dxdt.squeeze()

class FourierChebyshevFieldOpt(IOAble):
    _static_attrs = ["L", "M", "N", "M_fft", "N_fft", "data_keys"]
    def __init__(self, L, M, N):
        self.L = L
        self.M = M
        self.N = N

    def build(self, eq):
        self.data_keys = ["b", "grad(|B|)", "e^rho", "e^theta*rho", "e^zeta", "|B|"]
        self.l = jnp.arange(self.L)
        self.M_fft = 2 * self.M + 1
        self.N_fft = 2 * self.N + 1
        self.m = jnp.fft.fftfreq(self.M_fft) * self.M_fft
        self.n = jnp.fft.fftfreq(self.N_fft) * self.N_fft
        x = jnp.cos(jnp.pi * (2 * self.l + 1) / (2 * self.L))
        rho = (x + 1) / 2
        self.grid = LinearGrid(rho=rho, M=self.M, N=self.N, sym=False, NFP=1)
        self.transforms = get_transforms(self.data_keys, eq, self.grid, jitable=True)

    def fit(self, params, profiles):
        data_raw = compute_fun(
            "desc.equilibrium.equilibrium.Equilibrium",
            self.data_keys,
            params,
            self.transforms,
            profiles,
        )
        L, M, N = self.L, self.M_fft, self.N_fft
        bxgb = jnp.cross(data_raw["b"], data_raw["grad(|B|)"])

        betr = dot(data_raw["b"], data_raw["e^theta*rho"])
        bez = dot(data_raw["b"], data_raw["e^zeta"])
        bxgbr = dot(bxgb, data_raw["e^rho"])
        bxgbtr = dot(bxgb, data_raw["e^theta*rho"])
        bxgbz = dot(bxgb, data_raw["e^zeta"])
        bdgb = dot(data_raw["b"], data_raw["grad(|B|)"])
        B = data_raw["|B|"]

        keys = ["betr", "bez", "bxgbr", "bxgbtr", "bxgbz", "bdgb", "B"]
        # stack data to perform 7 transforms in batch
        stacked_data = jnp.stack(
            [
                # ber.reshape(N, L, M),
                betr.reshape(N, L, M),
                bez.reshape(N, L, M),
                bxgbr.reshape(N, L, M),
                bxgbtr.reshape(N, L, M),
                bxgbz.reshape(N, L, M),
                bdgb.reshape(N, L, M),
                B.reshape(N, L, M),
            ]
        )
        coefs = jax.scipy.fft.dct(stacked_data, axis=2, norm=None)
        # handle the 0-th Chebyshev coefficient and normalization
        coefs = coefs.at[:, :, 0, :].divide(2)
        coefs /= self.L

        coefs = jnp.fft.fft(coefs, axis=3, norm=None)
        coefs = jnp.fft.fft(coefs, axis=1, norm=None)

        data = {}
        coefs_real = coefs.real
        coefs_imag = coefs.imag

        for i, key in enumerate(keys):
            data[key + "_real"] = coefs_real[i]
            data[key + "_imag"] = coefs_imag[i]

        data["l"] = self.l
        data["m"] = self.m
        data["n"] = self.n
        data["M"] = self.M_fft
        data["N"] = self.N_fft

        self.params_dict = data

    def evaluate(self, rho, theta, zeta, params=None):
        if params is None:
            params = self.params_dict

        # the cosine transforms reverses the order
        r0p = 1 - 2 * rho
        Tl = jnp.cos(params["l"] * jnp.arccos(r0p))
        m_theta = params["m"] * theta
        expm_real = jnp.cos(m_theta) / params["M"]
        expm_imag = jnp.sin(m_theta) / params["M"]

        n_zeta = params["n"] * zeta
        expn_real = jnp.cos(n_zeta) / params["N"]
        expn_imag = jnp.sin(n_zeta) / params["N"]

        # The new shape for these arrays will be (k, n, l, m) where k=15
        cf_real_all = jnp.stack(
            [
                params[key + "_real"]
                for key in [
                    "betr",
                    "bez",
                    "bxgbr",
                    "bxgbtr",
                    "bxgbz",
                    "bdgb",
                    "B",
                ]
            ]
        )
        cf_imag_all = jnp.stack(
            [
                params[key + "_imag"]
                for key in [
                    "betr",
                    "bez",
                    "bxgbr",
                    "bxgbtr",
                    "bxgbz",
                    "bdgb",
                    "B",
                ]
            ]
        )

        # "knlm,l->knm" contracts the 'l' dimension for all 'k' batches at once
        f_l_real = jnp.einsum("knlm,l->knm", cf_real_all, Tl)
        f_l_imag = jnp.einsum("knlm,l->knm", cf_imag_all, Tl)

        # "knm,m->kn" contracts the 'm' dimension for all 'k' batches
        f_lm_real = jnp.einsum("knm,m->kn", f_l_real, expm_real) - jnp.einsum(
            "knm,m->kn", f_l_imag, expm_imag
        )
        f_lm_imag = jnp.einsum("knm,m->kn", f_l_real, expm_imag) + jnp.einsum(
            "knm,m->kn", f_l_imag, expm_real
        )

        # "kn,n->k" contracts the 'n' dimension, leaving just the batch dimension
        results = jnp.einsum("kn,n->k", f_lm_real, expn_real) - jnp.einsum(
            "kn,n->k", f_lm_imag, expn_imag
        )

        out = {}

        out["b*e^theta*rho"] = results[0]
        out["b*e^zeta"] = results[1]
        out["(bxgrad(|B|))*e^rho"] = results[2]
        out["(bxgrad(|B|))*e^theta*rho"] = results[3]
        out["(bxgrad(|B|))*e^zeta"] = results[4]
        out["b*grad(|B|)"] = results[5]
        out["|B|"] = results[6]

        return out
```
</details>

**Comment by YigitElma** [comment]
For some reason, this suffers from GPU synchronisation miserably, and for more particles, it is slower than direct approach.
